libp2p-swarm/src/swarm.rs

```diff
@@ -0,0 +1,226 @@
+          C: ConnectionUpgrade<T::RawConn> + Clone + 'static,      // TODO: 'static :-/
+		  C::NamesIter: Clone, // TODO: not elegant
+{
+    /// Asks the swarm to dial the node with the given multiaddress.
+    ///
+    /// Once the connection has been open and upgraded, it will be given to the handler.
+    // TODO: consider returning a future so that errors can be processed?
+    pub fn dial(&self, multiaddr: Multiaddr) -> Result<(), Multiaddr> {
+        match self.upgraded.clone().dial(multiaddr.clone()) {
+            Ok(dial) => {
+                // Ignoring errors if the receiver has been closed, because in that situation
+                // nothing is going to be processed anyway.
+                let _ = self.new_dialers.unbounded_send((dial, multiaddr));
+                Ok(())
+            },
+            Err((_, multiaddr)) => {
+                Err(multiaddr)
+            },
+        }
+    }
+
+    /// Adds a multiaddr to listen on.
+    pub fn listen_on(&self, multiaddr: Multiaddr) -> Result<Multiaddr, Multiaddr> {
+        match self.upgraded.clone().listen_on(multiaddr) {
+            Ok((listener, new_addr)) => {
+                // Ignoring errors if the receiver has been closed, because in that situation
+                // nothing is going to be processed anyway.
+                let _ = self.new_listeners.unbounded_send(listener);
+                Ok(new_addr)
+            },
+            Err((_, multiaddr)) => {
+                Err(multiaddr)
+            },
+        }
+    }
+}
+
+/// Future that must be driven to completion in order for the swarm to work.
+pub struct SwarmFuture<T, C, H, F>
+    where T: MuxedTransport + 'static,      // TODO: 'static :-/
+          C: ConnectionUpgrade<T::RawConn> + 'static,      // TODO: 'static :-/
+{
+    upgraded: UpgradedNode<T, C>,
+    handler: H,
+    new_listeners: mpsc::UnboundedReceiver<Box<Stream<Item = (Box<Future<Item = C::Output, Error = IoError>>, Multiaddr), Error = IoError>>>,
+    next_incoming: Box<Future<Item = (C::Output, Multiaddr), Error = IoError>>,
+    listeners: Vec<Box<Stream<Item = (Box<Future<Item = C::Output, Error = IoError>>, Multiaddr), Error = IoError>>>,
+    listeners_upgrade: Vec<(Box<Future<Item = C::Output, Error = IoError>>, Multiaddr)>,
+    dialers: Vec<(Box<Future<Item = C::Output, Error = IoError>>, Multiaddr)>,
+    new_dialers: mpsc::UnboundedReceiver<(Box<Future<Item = C::Output, Error = IoError>>, Multiaddr)>,
+    to_process: Vec<F>,
+}
+
+impl<T, C, H, If, F> Future for SwarmFuture<T, C, H, F>
+    where T: MuxedTransport + Clone + 'static,      // TODO: 'static :-/,
+          C: ConnectionUpgrade<T::RawConn> + Clone + 'static,      // TODO: 'static :-/
+          H: FnMut(C::Output, Multiaddr) -> If,
+          If: IntoFuture<Future = F, Item = (), Error = IoError>,
+          F: Future<Item = (), Error = IoError>,
+{
+    type Item = ();
+    type Error = IoError;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
```

Comment:
This is a massively long function, is there any way to break it up in more comprehensible pieces? It's not like it's super hard to understand, it's just the same pattern over and over again, but I'd be inclined to organize it into smaller functions.

Reply:
I went with a long function like this, because in my opinion it would be *more* confusing to use separate functions or macros. But I can give it a shot if you think it would be better.

Reply:
Yeah, it's hard to say. You may be right, _because_ it is a simple pattern just repeating. I don't have a strong opinion on it. Keep it if you think this is the cleanest way.